### PR TITLE
Finish implementing GATs

### LINKED ISSUE #116: Implement GATs
## Step 1
The grammar needs to be extended to support bounds and (quantified) where clauses on associated types like in:
```rust
trait Foo {
    type Item<'a, T>: Clone where Self: Sized;
}
```

The grammar is located [parser.lalrpop](https://github.com/rust-lang-nursery/chalk/blob/master/chalk-parse/src/parser.lalrpop), and the AST in [ast.rs](https://github.com/rust-lang-nursery/chalk/blob/master/chalk-parse/src/ast.rs).

When this is done, this item:
https://github.com/rust-lang-nursery/chalk/blob/eeb21829c1e847921261ad820b5b2ec35b649c76/chalk-parse/src/parser.lalrpop#L204-L207
won't be needed anymore because we will be using `QuantifiedWhereClauses` everywhere.

We would also remove where clauses on associated type values because they are useless:
https://github.com/rust-lang-nursery/chalk/blob/eeb21829c1e847921261ad820b5b2ec35b649c76/chalk-parse/src/parser.lalrpop#L104-L111

## Step 2
Currently, a where clause of the form `where T: Foo<Item = U>` translated to a single domain goal `ProjectionEq(<T as Foo>::Item = U)`. We would need this clause to be translated to two distinct domain goals:
* `ProjectionEq(<T as Foo>::Item = U)` as before
* `Implemented(T: Foo)`

Currently, we cannot translate one AST where clause to multiple IR domain goals. This means that the following trait:
https://github.com/rust-lang-nursery/chalk/blob/eeb21829c1e847921261ad820b5b2ec35b649c76/src/lower/mod.rs#L437-L439
should be changed into something like:
```rust
trait LowerWhereClause<T> {
    fn lower(&self, env: &Env) -> Result<Vec<T>>;
}
```

## Step 3
Now we should be in a position where we can implement the various rules written in [this comment](https://github.com/rust-lang-nursery/chalk/issues/12#issuecomment-377258347) (niko's comment just below can be ignored, it was basically merged into the main one). This includes:
* program clauses which should be implemented in [lower/mod.rs](https://github.com/rust-lang-nursery/chalk/blob/master/src/lower/mod.rs)
* WF requirements which should be implemented in [lower/wf.rs](https://github.com/rust-lang-nursery/chalk/blob/master/src/lower/wf.rs)

Note: the organization of these two files will change after #114 is fixed, so the preceding text may become obsolete.

## Step 4
At that point, the domain goal `WellFormed(ProjectionEq)` will be unused. See #115 for the next steps.

cc @rust-lang-nursery/wg-traits

### PR DESCRIPTION
**Pull request by tmandry**
Implement and test logic around Generic Associated Types.

This required shuffling around the logic rules we used. This PR implements the set of rules described in https://github.com/rust-lang-nursery/chalk/issues/12#issuecomment-377258347.

Closes #116.

### CODE REVIEW

**Comment by tmandry** [comment]
Ready for review! Various disclaimers..

This doesn't yet support higher-ranked inline bounds.

The following clause is not tested:
```
FromEnv(Self: Foo) :- FromEnv((Foo::Item<'a,T>)<Self>).
```
I'm not sure it's possible to test this in the current grammar.

I just remembered I want to clean up some repetition in `lowering.rs` later.

File: src/rules/wf.rs [hunk 1]
```diff
@@ -256,16 +240,50 @@ impl WfSolver {
...
+                      .flat_map(|b| b.lower_with_self(assoc_ty.value.value.ty.clone()))
+                      .map(|g| g.into_well_formed_goal());
+            
+            let goals = wf_goals.chain(bound_goals).casted();
+            let goal = match goals.fold1(|goal, leaf| Goal::And(Box::new(goal), Box::new(leaf))) {
+                Some(goal) => goal,
+                None => return None,
+            };
+
+            let hypotheses =
```
Comment (scalexm) [inline src/rules/wf.rs]: and saying that we do the same for where clauses here, regarding the comment above

File: src/rules/wf/test.rs [hunk 2]
```diff
@@ -217,7 +217,7 @@ fn projection_type_in_header() {
 
             // Projection types in an impl header are not assumed to be well-formed,
             // an explicit where clause is needed (see below).
```
Comment (scalexm) [inline src/rules/wf/test.rs]: The comment is obsolete and can be removed.

File: src/solve/test.rs [hunk 3]
```diff
@@ -567,6 +760,46 @@ fn normalize_implied_bound() {
     }
 }
 
+#[test]
+fn normalize_implied_bound_gat() {
```
Comment (scalexm) [inline src/solve/test.rs]: Pre-existing: could this be renamed to just "gat_implied_bounds" or something like this?

File: src/rules/wf/test.rs (hunk 2 above)
Comment (tmandry) [reply src/rules/wf/test.rs]: I think it's still mostly correct, but I changed the wording some.

**Pull request merged**